rpc/src/main/scala/org/bitcoins/rpc/serializers/JsonReaders.scala

```diff
@@ -0,0 +1,313 @@
+
+import java.io.File
+import java.net.{InetAddress, URI}
+
+import org.bitcoins.core.crypto.{
+  DoubleSha256Digest,
+  ECPublicKey,
+  Sha256Hash160Digest
+}
+import org.bitcoins.core.currency.{Bitcoins, Satoshis}
+import org.bitcoins.core.number.{Int32, Int64, UInt32, UInt64}
+import org.bitcoins.core.protocol.{
+  Address,
+  BitcoinAddress,
+  P2PKHAddress,
+  P2SHAddress
+}
+import org.bitcoins.core.protocol.blockchain.{Block, BlockHeader, MerkleBlock}
+import org.bitcoins.core.protocol.script.{ScriptPubKey, ScriptSignature}
+import org.bitcoins.core.protocol.transaction._
+import org.bitcoins.core.wallet.fee.{BitcoinFeeUnit, SatoshisPerByte}
+import org.bitcoins.rpc.jsonmodels.RpcAddress
+import play.api.libs.json._
+
+import scala.util.{Failure, Success, Try}
+
+object JsonReaders {
+  private def processJsNumber[T](numFunc: BigDecimal => T)(
+      json: JsValue): JsResult[T] = json match {
+    case JsNumber(n) => JsSuccess(numFunc(n))
+    case err @ (JsNull | _: JsBoolean | _: JsString | _: JsArray |
+        _: JsObject) =>
+      JsError(s"error.expected.jsnumber, got ${Json.toJson(err).toString()}")
+  }
+
+  private def processJsString[T](strFunc: String => T)(
+      json: JsValue): JsResult[T] = json match {
+    case JsString(s) => JsSuccess(strFunc(s))
+    case err @ (JsNull | _: JsBoolean | _: JsNumber | _: JsArray |
+        _: JsObject) =>
+      JsError(s"error.expected.jsstring, got ${Json.toJson(err).toString()}")
+  }
+
+  implicit object BigIntReads extends Reads[BigInt] {
+    override def reads(json: JsValue): JsResult[BigInt] =
+      processJsNumber[BigInt](_.toBigInt())(json)
+  }
+
+  implicit object DoubleSha256DigestReads extends Reads[DoubleSha256Digest] {
+    override def reads(json: JsValue): JsResult[DoubleSha256Digest] =
+      processJsString[DoubleSha256Digest](DoubleSha256Digest.fromHex)(json)
+  }
+
+  implicit object BitcoinsReads extends Reads[Bitcoins] {
+    override def reads(json: JsValue): JsResult[Bitcoins] =
+      processJsNumber[Bitcoins](Bitcoins(_))(json)
+  }
+
+  implicit object SatoshisReads extends Reads[Satoshis] {
+    override def reads(json: JsValue): JsResult[Satoshis] =
+      processJsNumber[Satoshis](num => Satoshis(Int64(num.toBigInt)))(json)
+  }
+
+  implicit object BlockHeaderReads extends Reads[BlockHeader] {
+    override def reads(json: JsValue): JsResult[BlockHeader] =
+      processJsString[BlockHeader](BlockHeader.fromHex)(json)
+  }
+
+  implicit object Int32Reads extends Reads[Int32] {
+    override def reads(json: JsValue): JsResult[Int32] = json match {
+      case JsNumber(n) =>
+        n.toBigIntExact() match {
+          case Some(num) => JsSuccess(Int32(num))
+          case None      => JsError(s"error.expected.Int32, got $n")
+        }
+      case JsString(s) => JsSuccess(Int32.fromHex(s))
+      case err @ (JsNull | _: JsBoolean | _: JsArray | _: JsObject) =>
+        JsError(s"error.expected.jsnumber, got ${Json.toJson(err).toString()}")
```

Comment:
Let's create a `buildErrorMsg` method to do this stuff instead of being redundant